### PR TITLE
Python SDK - Phase 1

### PR DESCRIPTION
**Pull request by cretz**
# Summary

This is an initial proposal to start development on the Python SDK.

[See it rendered here](https://github.com/cretz/temporal-proposals/blob/python-phase1/python/phase-1.md) (may be deleted when proposal is done). Also see [the previous draft PR here](https://github.com/temporalio/proposals/pull/49) where several discussions were had.

**NOTE: This is not an in-depth proposal w/ accurate APIs and such, but rather a simple proposal to reach alignment before work begins.**

Feel free to leave a bunch of comments. If discussion becomes too difficult to follow, I will close this PR and start a new one (like I already have once :-)).

### CODE REVIEW

File: python/phase-1.md [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+  * Probably
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
+exported from `temporalio`):
+
+```python
+class Client:
+    def __init__(core_or_service: temporalio.Core) -> None:
```
Comment (cretz) [inline python/phase-1.md:120]:
```suggestion
    def __init__(core: temporalio.Core) -> None:
```

(leftover from an old version that accepted a union of core or pure Python gRPC which we don't want atm)

File: python/phase-1.md [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+        # internally.
+        activity_executor: Optional[concurrent.futures.Executor]=None,
+        # Whether to use the given/defaulted executor for async activities as
+        # well.
+        use_activity_executor_on_async: bool = False,
+        # ...
+    ) -> None:
+        pass
+
+    async def run() -> None:
```
Comment (cretz) [inline python/phase-1.md:244]: Missing `self`

File: python/phase-1.md [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
+exported from `temporalio`):
+
+```python
+class Client:
+    def __init__(core_or_service: temporalio.Core) -> None:
+        pass
```
Comment (Sushisource) [inline python/phase-1.md:121]: I don't think we want to init clients this way - I think `Core` should have a `get_client` method, and I think a client constructor should accept some client options object (which Core will also accept when created) which it will use to create a core-independent client via the publicly exposed interface here: https://github.com/temporalio/sdk-core/blob/e3f54276f971bcc9b6e1d1cf4d0ecbeb853387ae/client/src/lib.rs#L210

IE: You should be able to have a client tied to core if you want, but you should also be able to get a client without any core at all.

Obviously the rub is telemetry like we talked about a little bit. We can clean up how that works in core to make it easy, but we can start without any changes by simply having clients-not-from-core not do any telemetry.

File: python/phase-1.md [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+            "activity1_method": MyActivities().activity1,
+            # Can use lambda
+            "lambda": lambda arg1: arg1,
+            # Fails because not callable
+            "not_callable": MyActivities(),
+        },
+    )
+```
+
+Notes:
```
Comment (cretz) [inline python/phase-1.md:375]: Need to document our discussions around why we removed activity decorator and why we require a mapping w/ explicit activity name

File: python/phase-1.md (hunk 3 above)
Comment (cretz) [reply python/phase-1.md:121]: I had originally had similar, but:

* I didn't want to make an implementation of this class with non-core because I couldn't figure out how to return `service` properly as I don't think the core and non-core will have the same signatures for the call. So I am not sure I can have a core-independent client for all gRPC calls unless I maintain all gRPC calls manually. But I haven't dug enough to confirm.
* We don't have `Core.create_worker`, so I figured it was more consistent to pass core to clients/workers than to either be inconsistent or force worker creation to be on core too.

I personally think it's cleaner this way (and any other potential uses of the same core instance could potentially take core in their constructor too). But I can definitely change this if consensus is to do so.

Not too worried about the telemetry stuff, we can figure that out a bit later (I have some ideas that basically moves `tokio::tracing` to lang-side and requires core `init` to accept a spawner unfortunately since I can't see how to do thread-local on tokio runtime threads, but that's for a different setting).

File: python/phase-1.md [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+
+* After discussion, it has been decided that both clients and workers will be backed by core
+    * Yes that means clients require the Rust library which we accept
+* After discussion, it has been decided that we will not have an implicit local default like TypeScript SDK, but instead
+  an explicit default
+  * `default_local()` memoizes success
+* Should we have a constructor, an `__aenter__`, and an `__aexit__`?
+  * Probably
```
Comment (bergundy) [inline python/phase-1.md:111]: yes please, that's very pythonic.

File: python/phase-1.md (hunk 3 above)
Comment (bergundy) [reply python/phase-1.md:121]: My main issue with tying Core to the Client is that they're 2 completely different entities and it's actually Core that takes a client and not the other way around on the rust side).

File: python/phase-1.md [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+        client: temporalio.Client,
+        workflow_id: str,
+        *,
+        run_id: Optional[str] = None
+        # ...
+    ) -> None:
+        pass
+
+    def __await__(self) -> T:
+        pass
```
Comment (bergundy) [inline python/phase-1.md:179]: Another issue with this approach is that `await client.get_workflow_handle()` now waits will wait for completion which could be confusing.

Comment (bergundy) [reply python/phase-1.md:179]: Let's reconsider adding an explicit `result` method.

File: python/phase-1.md [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+* We have left off workflow registration until a later phase
+* Stopping the worker is done via cancellation of the run task/future
+* The constructor will fail if the task queue has already been registered with that instance of core
+* Internally, activities will run in the same event loop `run()` is called on
+* We allow a custom executor and potentially even default one for sync activities
+  * Should we default an executor for sync?
+  * Is it ok to not default an executor for async? Or should we use it for them too?
+* Should we have an `__aenter__` and an `__aexit__`?
+  * Probably, and change the examples to use `with` to ensure worker is shut down properly
+  * This means we have to treat those two calls as separate start and shutdown
```
Comment (bergundy) [inline python/phase-1.md:259]: Doesn't look like this would work here.
What would be in the body of the `with` statement? It'd have to block on something.

File: python/phase-1.md [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+
+Notes:
+
+* We have left off workflow registration until a later phase
+* Stopping the worker is done via cancellation of the run task/future
+* The constructor will fail if the task queue has already been registered with that instance of core
+* Internally, activities will run in the same event loop `run()` is called on
+* We allow a custom executor and potentially even default one for sync activities
+  * Should we default an executor for sync?
+  * Is it ok to not default an executor for async? Or should we use it for them too?
```
Comment (bergundy) [inline python/phase-1.md:256]: I vote against default executors, explicit is better than implicit here too IMHO.
(I've raised my concerns in the previous PR)

File: python/phase-1.md [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+    )
+```
+
+Notes:
+
+* Cancellation is task cancellation
+  * We need to research whether, in Python 3.7, we can interrupt an async task w/ our own error reliably
+* Heartbeat is a synchronous call
+  * Should we throw a cancellation error if we know the activity is cancelled when they try to heartbeat?
+    * is this something core should be returning from `record_heartbeat` instead of lang?
```
Comment (bergundy) [inline python/phase-1.md:381]: heartbeat should not be tied to cancellation, this separation allows local activities to be cancelled.

File: python/phase-1.md [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+        pass
+
+    async def encode(self, *values):
+        pass
+
+    async def decode(self, *payloads):
+        pass
+
+
+PayloadConverter.default = CompositePayloadConverter(
```
Comment (bergundy) [inline python/phase-1.md:418]: After thinking about this some more should Python support [pickle](https://docs.python.org/3/library/pickle.html) by default?
A lot of python devs would expect us to support this (for arbitrary object serialization).
One obvious drawback to this is that it's not a polyglot solution.

File: python/phase-1.md (hunk 3 above)
Comment (Sushisource) [reply python/phase-1.md:121]: I suppose it works (to use only a client in your python process) as long as you never register any workers. It's not really how I've thought of what Core is, but I guess it technically works. It's not super clean. You still have this core instance that you can "shut down" but that doesn't really mean anything, and it's got facilities for things you're never going to use if all you want is a client.

And "I couldn't figure out how to return service properly as I don't think the core and non-core will have the same signatures for the call. " is easily solved. It's not exposed right now in core but it's trivial to add, so I wouldn't let that stop you.

Regarding "Core.create_worker" -- why not? I mean I don't care a whole lot either way here but I think that makes sense too. If we want it to accept core that's fine still, and client can either accept core (`Client.from_core`), or can accept options for a core-independent client.

File: python/phase-1.md (hunk 10 above)
Comment (Sushisource) [reply python/phase-1.md:418]: I'm not really sure why you'd need it - unless maybe you're passing already-pickled objects from some datastore directly into activities/wfs (which I can believe).

Otherwise, if you're using our client or package, you're going to get a reasonable form of serialization for free already (with one of the defaults), so why do you care if it's pickle?

At minimum, seems like something we can add later if wanted, doesn't need to be in the proposal.

File: python/phase-1.md (hunk 3 above)
Comment (cretz) [reply python/phase-1.md:121]: > My main issue with tying Core to the Client is that they're 2 completely different entities and it's actually Core that takes a client and not the other way around on the rust side).

Can you clarify this? Are you talking about the off-PR discussion of whether to tie the client to core? I had understood that as settled.

> And "I couldn't figure out how to return service properly as I don't think the core and non-core will have the same signatures for the call. " is easily solved. It's not exposed right now in the rust client crate but it's trivial to add, so I wouldn't let that stop you.

I am not sure it's easily solved. If the Python gRPC client, say, takes contextual call options I can't expect to abstract those properly.

> Regarding "Core.create_worker" -- why not? I mean I don't care a whole lot either way here but I think that makes sense too. If we want it to accept core that's fine still, and client can either accept core (Client.from_core), or can accept options for a core-independent client.

I think it's cleaner to create instances of classes via Python's class creation mechanisms whenever reasonable.

Can I get clarity on what's being asked for here? So we want `get_client` and `create_worker` on core? Or we don't want tying Core to a client? Or we want an abstract client that works on pure Python gRPC and Core client including abstracting all workflowservice calls over both? I am looking for concrete complete suggestions (the piece here, even down to only have a Core version and to have the client and worker accept core in constructor non-defaulted was what was settled on during discussion, but maybe it was unclear and of course we can change).

File: python/phase-1.md (hunk 6 above)
Comment (cretz) [reply python/phase-1.md:179]: > Another issue with this approach is that await client.get_workflow_handle() now waits will wait for completion which could be confusing.

This is only unclear because I was forced to change the name. `await client.workflow("my workflow")` was very clear originally from the previous PR.

File: python/phase-1.md (hunk 8 above)
Comment (cretz) [reply python/phase-1.md:256]: Agreed, so we won't use a default executor for sync.

The question here becomes, and sorry it was worded unclearly, do we want async functions to use the same executor as sync functions? Or do we want a bool like I have here? Or do we want to allow a separate executor to be passed for async functions?

File: python/phase-1.md (hunk 9 above)
Comment (cretz) [reply python/phase-1.md:381]: Ok

File: python/phase-1.md (hunk 10 above)
Comment (cretz) [reply python/phase-1.md:418]: >  One obvious drawback to this is that it's not a polyglot solution.

The most obvious. But I'll provide a pickle-based converter they can choose to use if they want.

File: python/phase-1.md (hunk 3 above)
Comment (cretz) [reply python/phase-1.md:121]: I am marking this resolved as we are having off-PR discussions about how core, clients, and workers interact.

File: python/phase-1.md (hunk 6 above)
Comment (bergundy) [reply python/phase-1.md:179]: I don't think the name change is what's making it unclear.
It's that you get 2 completely different results with an without `await`.

Comment (cretz) [reply python/phase-1.md:179]: That's normal I think for awaitable objects that carry other info. I will change to adding a result method though.

File: python/phase-1.md (hunk 8 above)
Comment (bergundy) [reply python/phase-1.md:256]: I don't think we should mess with executors at all.
We should just run async functions (function that returns a Future) on the currently installed event loop in a single threaded fashion.
I'd start with async only activities and explain that if users want to run sync code they need to wrap it in an executor.

File: python/phase-1.md (hunk 10 above)
Comment (bergundy) [reply python/phase-1.md:418]: The problem with JSON is that by default python objects aren't serializable, only primitives are.
We can recommend using dataclasses which can be turned into dicts https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict

File: python/phase-1.md (hunk 1 above)
Comment (cretz) [reply python/phase-1.md]: Fixed.

File: python/phase-1.md (hunk 2 above)
Comment (cretz) [reply python/phase-1.md]: Fixed

File: python/phase-1.md (hunk 4 above)
Comment (cretz) [reply python/phase-1.md:383]: Documented

File: python/phase-1.md (hunk 5 above)
Comment (cretz) [reply python/phase-1.md]: Done

File: python/phase-1.md (hunk 6 above)
Comment (cretz) [reply python/phase-1.md]: Added `result()`

File: python/phase-1.md (hunk 7 above)
Comment (cretz) [reply python/phase-1.md]: Added `start` and `shutdown`

File: python/phase-1.md (hunk 8 above)
Comment (cretz) [reply python/phase-1.md]: I think that's a bit much. I think it's normal to only put `async` on a function if you need to. Also, there could be value in running async functions in executor too. But yes, I accept both as params that are defaulted to none atm.

File: python/phase-1.md (hunk 9 above)
Comment (cretz) [reply python/phase-1.md]: Updated

File: python/phase-1.md (hunk 10 above)
Comment (cretz) [reply python/phase-1.md:435]: > The problem with JSON is that by default python objects aren't serializable, only primitives are.

In all SDKs we accept the limitation that the in/out types must be serializable by the converter by default. What happens if I have a function on a TypeScript type? Technically they could subclass `json.JSONEncoder` and impl `default` for out types, but yeah it'd just come back as the raw form of that.

>  We can recommend using dataclasses for args and return types which can be turned into dicts https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict

We can do this for "encode" of the `JSONPlainPayloadConverter`, but decoded values will remain dicts. So our suggestion should really be "use dicts directly or dataclass knowing it will come in as a dict that you'll have to create your class with if needed". But I think this limitation is fine compared to, say, trying to store the class name in the payload by default (but we can definitely support that if needed, same as we support pickle but not by default).

**Comment by roselander** [comment]
In my previous life (not very long ago) I maintained a library that used Python's asyncio, "await" and I thought it was extremely half-baked.

Crucially, the only methods which truly implement async behavior are low-level I/O primitives such as opening a TCP socket. In other words, unless you eventually delegate to one of these "magic" methods, your "async" code will never release the Global Interpreter Lock, which kind of defeats the purpose IMO.

So basically my question is whether someone who understands Python asyncio at a deep level has thought about how this would actually be implemented. I'm kind of assuming that's the case but just wanted to double-check.

**Comment by bergundy** [comment]
@roselander I actually think asyncio is pretty well designed, Guido was the one who lead the effort and he based it on popular async libraries like twisted and tornado.

Coroutines have nothing to do with the GIL, they help you write concurrent code in a single threaded environment.
If you want to use multiple cores in python your only option is to use multiple processes or use C extensions which can release the GIL.
Our Python SDK, is going to use asyncio and be an async first SDK for a couple of reasons:
- You can use synchronous code in an async environment and not the other way around (e.g. the async keyword is contagious).
- For workflow logic we want to be able to reuse known concurrency primitives such as [`gather`](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather) and [`as_completed`](https://docs.python.org/3/library/asyncio-task.html#asyncio.as_completed)

File: python/phase-1.md [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+
+* For demonstration/simplicity here we are leaving off a lot of proper typings (e.g. `Optional`) that the real version
+  will have
+* We leave off a lot of options on calls for this high-level proposal
+* We are not using proper function documentation, but rather inline comments for easier discussion
+* **NOTE: These will have proper docs and types when developed!**
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
```
Comment (cretz) [inline python/phase-1.md:80]: I am thinking this may not be clearest exported from `temporalio` directly, but instead leave only in the `client` module. So you would have to `from temporalio import client` and use `client.Client`. This more aligns with our expectations of using `from temporalio import activity` for e.g. `activity.info()`. Thought I do think all common code such as retries and errors may deserve to be in the `temporalio` package directly.

Thoughts?

File: python/phase-1.md [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+        self, *values: Any
+    ) -> Union[temporalio.Payload, list[temporalio.Payload]]:
+        pass
+
+    # This can be sync or async. Always takes a list. Can return a single value
+    # or a list of values with the same size as the input list or None if the
+    # value cannot be converted with this converter. Any other result causes an
+    # error. Any error raised is bubbled.
+    @abstractmethod
+    async def decode(self, *payloads: temporalio.Payload) -> Union[Any, list[Any]]:
```
Comment (cretz) [inline python/phase-1.md:422]: There is an (obvious in hindsight) ambiguity issue here. What if your single payload is a list?

Any suggestions on how to combine "decode single payload -> multiple values" and "decode multiple payloads -> multiple values" and "decode multiple payloads -> single value"? Do we have to support the last use case?

I am now thinking about adding extra `encode_multiple` and `decode_multiple` calls that default to just calling these and if someone, say for encryption purposes, wants to convert multiple to single, they have to override those two also. Thoughts?

Comment (cretz) [reply python/phase-1.md:422]: Due to additional ambiguities, I'm gonna have to return a `Tuple[Any, bool]` from `decode(self, payload)` with the second value being `True` if it was valid. There is ambiguity if I just return `-> Any` because what if `None` is a real value?

Another approach would be to separate out into data and payload converters again, but that seems unnecesary.

**Comment by cretz** [comment]
@roselander - While this API will be available because it is autogenerated from our API, we're actually going to be delegating to a native extension in Rust to make gRPC invocations since that extension is already handling advanced server communication and event handling for the worker. This is a deviation from TypeScript which has an in-lang gRPC client separate from the Rust client. After discussion of the tradeoffs of separate connections vs requiring a C extension (in Rust), we settled on the latter.

**Comment by roselander** [comment]
@cretz 

> Can you clarify why this is a problem? Our extension will end up calling asyncio functions as well.

There was no asyncio gRPC at the time, and so in practice it wasn't actually possible for us to implement our async library, because there was no existing async primitive to build our library on top of.

File: python/phase-1.md [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+
+    async def shutdown(self) -> None:
+        pass
+
+    async def __aenter__(self) -> Worker:
+        await self.start()
+        return self
+
+    async def __aexit__(self) -> None:
+        await self.shutdown()
```
Comment (bergundy) [inline python/phase-1.md:252]: I don't think this is required, what is this supposed to block on?

```
async with worker:
  # what goes here?
```

File: python/phase-1.md [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+
+```python
+class PayloadConverter(ABC):
+    # This can be sync or async. Always takes a list. Can return a single
+    # payload or a list of payloads with the same size as the input list or None
+    # if the value cannot be converted with this converter. Any other result
+    # causes an error. Any error raised is bubbled.
+    @abstractmethod
+    async def encode(
+        self, *values: Any
```
Comment (bergundy) [inline python/phase-1.md:413]: After thinking about this a bit, we might want to use reflection here to get the expected type from the activity signature like we do in Java / Go

File: python/phase-1.md (hunk 12 above)
Comment (bergundy) [reply python/phase-1.md:422]: I thought these methods were meant for multiple payloads, it's okay to have a payload that's an array.
The method can map M to N payloads.

**Review by bergundy — APPROVED** [review]
There are still 1-2 unresolved items but I'm already approving

File: python/phase-1.md (hunk 13 above)
Comment (cretz) [reply python/phase-1.md:252]:
```
async with worker:
  # everything you want to do with a worker
```

If you want to use it that way. The point is that it stops when done there. If you don't have your code encapsulated like that, you can just run the worker, you don't have to use `async with`.

File: python/phase-1.md (hunk 14 above)
Comment (cretz) [reply python/phase-1.md:413]: I agree. I can mutate this proposal to include that (of course not require type annotations). I will have to do some research on runtime annotation resolution.

**Comment by cretz** [comment]
I am merging this as most of our discussions are now to implementation specifics which we can do on PRs that are already rolling in to https://github.com/temporalio/sdk-python. The exact APIs of this phase, as predicted, are changing during implementation.
